Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,159 @@
+Note that a `LocalDiffeomorphAux` is not `FunLike` (like `LocalHomeomorph`),
+as `toFun` doesn't determine `invFun` outside of `target`. -/
+instance : CoeFun (LocalDiffeomorphAux I J M N n) fun _ => M → N :=
+  ⟨fun Φ => Φ.toFun'⟩
+
+/-- A diffeomorphism is a local diffeomorphism. -/
+def Diffeomorph.toLocalDiffeomorphAux (h : Diffeomorph I J M N n) : LocalDiffeomorphAux I J M N n :=
+  {
+    toLocalHomeomorph := h.toHomeomorph.toLocalHomeomorph
+    contMDiffOn_toFun := fun x _ ↦ h.contMDiff_toFun x
+    contMDiffOn_invFun := fun _ _ ↦ h.symm.contMDiffWithinAt
+  }
+
+-- Add the very basic API we need.
+namespace LocalDiffeomorphAux
+variable (Φ : LocalDiffeomorphAux I J M N n) (hn : 1 ≤ n)
+protected theorem contMDiffOn : ContMDiffOn I J n Φ Φ.source :=
+  Φ.contMDiffOn_toFun
+
+protected theorem contMDiffOn_symm : ContMDiffOn J I n Φ.invFun Φ.target :=
+  Φ.contMDiffOn_invFun
+
+protected theorem mdifferentiableOn : MDifferentiableOn I J Φ Φ.source :=
+  (Φ.contMDiffOn).mdifferentiableOn hn
+
+protected theorem mdifferentiableOn_symm : MDifferentiableOn J I Φ.invFun Φ.target :=
+  (Φ.contMDiffOn_symm).mdifferentiableOn hn
+
+protected theorem mdifferentiableAt {x : M} (hx : x ∈ Φ.source) : MDifferentiableAt I J Φ x :=
+  (Φ.mdifferentiableOn hn x hx).mdifferentiableAt (Φ.open_source.mem_nhds hx)
+
+-- define symm just to make this easier to write?
+protected theorem mdifferentiableAt_symm {x : M} (hx : x ∈ Φ.source) :
+    MDifferentiableAt J I Φ.invFun (Φ x) :=
+  (Φ.mdifferentiableOn_symm hn (Φ x) (Φ.map_source hx)).mdifferentiableAt
+  (Φ.open_target.mem_nhds (Φ.map_source hx))
+
+/- We could add lots of additional API (following `Diffeomorph` and `LocalHomeomorph*), such as
+* further continuity and differentiability lemmas
+* refl, symm and trans instances; lemmas between them.
+As this declaration is meant for internal use only, we keep it simple. -/
+end LocalDiffeomorphAux
+end LocalDiffeomorphAux
+
+variable {M N}
+
+/-- `f : M → N` is called a **`C^n` local diffeomorphism at *x*** iff there exist
+  open sets `U ∋ x` and `V ∋ f x` and a diffeomorphism `Φ : U → V` such that `f = Φ` on `U`. -/
+def IsLocalDiffeomorphAt (f : M → N) (x : M) : Prop :=
+  ∃ Φ : LocalDiffeomorphAux I J M N n, x ∈ Φ.source ∧ EqOn f Φ Φ.source
+
+/-- `f : M → N` is a **`C^n` local diffeomorphism** iff it is a local diffeomorphism
+at each `x ∈ M`. -/
+def IsLocalDiffeomorph (f : M → N) : Prop :=
+  ∀ x : M, IsLocalDiffeomorphAt I J n f x
+
+lemma isLocalDiffeomorph_iff {f : M → N} :
+    IsLocalDiffeomorph I J n f ↔ ∀ x : M, IsLocalDiffeomorphAt I J n f x := by rfl
+
+variable {n}
+
+/-- A `C^n` diffeomorphism is a local diffeomorphism. -/
+lemma Diffeomorph.isLocalDiffeomorph (Φ : M ≃ₘ^n⟮I, J⟯ N) : IsLocalDiffeomorph I J n Φ :=
+  fun _ ↦ ⟨Φ.toLocalDiffeomorphAux, by trivial, eqOn_refl Φ _⟩
+
+/-- The image of a local diffeomorphism is open. -/
+def LocalDiffeomorph.image {f : M → N} (hf : IsLocalDiffeomorph I J n f) : Opens N := by
```

Comment:
Please also have `isOpen (range f)` as a separate lemma.

Reply:
Makes sense - it's basically just extracting that proof. Added.